nixos/modules/services/matrix/mjolnir.nix

```diff
@@ -0,0 +1,239 @@
+{ config, lib, pkgs, ... }:
```

Comment:
The service is in a restart loop by default when connecting to an EMS hosted Matrix:

```
Jun 14 15:14:45 kif systemd[1]: Started mjolnir - a moderation tool for Matrix.
Jun 14 15:14:45 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:45 GMT [INFO] [index] Starting bot...
Jun 14 15:14:46 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:46 GMT [INFO] [index] Resolving protected rooms...
Jun 14 15:14:46 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:46 GMT [INFO] [index] Resolving management room...
Jun 14 15:14:47 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:47 GMT [ERROR] [MatrixHttpClient (REQ-4)] {
Jun 14 15:14:47 kif mjolnir[22947]:   consent_uri: 'https://xxx.ems.host/_matrix/consent?u=xxx&h=xxx',
Jun 14 15:14:47 kif mjolnir[22947]:   errcode: 'M_CONSENT_NOT_GIVEN',
Jun 14 15:14:47 kif mjolnir[22947]:   error: 'To continue using this homeserver you must review and agree to our terms and conditions at https://xxx.ems.host/_matrix/consent?u=xxx&h=xxx.'
Jun 14 15:14:47 kif mjolnir[22947]: }
Jun 14 15:14:47 kif mjolnir[22947]: (node:22947) UnhandledPromiseRejectionWarning: #<IncomingMessage>
Jun 14 15:14:47 kif mjolnir[22947]: (node:22947) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
Jun 14 15:14:47 kif mjolnir[22947]: (node:22947) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Jun 14 15:14:47 kif systemd[1]: mjolnir.service: Succeeded.
Jun 14 15:14:47 kif systemd[1]: mjolnir.service: Consumed 495ms CPU time, received 3.1K IP traffic, sent 1.9K IP traffic.
```

I'm a bit surprised it is "Succeeded" instead of failed here, but perhaps this is a bug in mjolnir?

Reply:
After consenting, I get another error that kills the bot but exits with success:

```
Jun 14 15:17:15 kif mjolnir[29663]: Mon, 14 Jun 2021 15:17:15 GMT [ERROR] [MatrixHttpClient (REQ-4)] { errcode: 'M_FORBIDDEN', error: 'You are not invited to this room.' }
Jun 14 15:17:15 kif mjolnir[29663]: (node:29663) UnhandledPromiseRejectionWarning: #<IncomingMessage>
Jun 14 15:17:15 kif mjolnir[29663]: (node:29663) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
Jun 14 15:17:15 kif mjolnir[29663]: (node:29663) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Reply:
Had something similar happen when starting up mjolnir and pantalaimon wasn't ready to serve requests yet. The mjolnir process exited with status 0 as well (https://github.com/NixOS/nixpkgs/pull/123896#discussion_r637407051).

Looking at the node warning however, it seems like the mjolnir code doesn't catch that promise rejection. Will try to identify where the check is missing and try to create a patch upstream.

Reply:
Added an upstream PR catching these errors and setting a non-zero exit code. Also added that change as a temporary patch to the mjolnir package.